src/exo/LoopIR_scheduling.py

```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
+        self.env.exit_scope()
+        return self.bounds.pop()
+
+    def validate_access(self, idx: Optional[int]) -> bool:
+        if self.bounds is None:
+            return
+        elif idx is None or self.bounds[-1].hi is None:
+            self.check_passed = False
+        else:
+            self.check_passed &= idx > self.bounds[-1].hi - self.size
+
+    def update_bounds(self, new_bounds: IndexRange, check_safety: bool = False):
+        if self.bounds[-1] is None:
+            self.bounds[-1] = new_bounds
+        else:
+            if check_safety:
+                self.validate_access(new_bounds.lo)
+            self.bounds[-1] |= new_bounds
+
+    def do_stmts(self, stmts):
+        self.enter_scope()
+        super().do_stmts(stmts)
+        return self.exit_scope()
+
+    def do_s(self, s):
+        bounds = None
+        if isinstance(s, LoopIR.For):
+            bounds = self.do_stmts(s.body)
+            env = {
+                s.iter: (
+                    s.lo.val if isinstance(s.lo, LoopIR.Const) else None,
+                    s.hi.val - 1 if isinstance(s.hi, LoopIR.Const) else None,
+                )
+            }
+            bounds = bounds.eval_base_with_env(env)
+        elif isinstance(s, LoopIR.If):
+            if_bounds = self.do_stmts(s.body)
+            orelse_bounds = self.do_stmts(s.orelse)
+            bounds = merge_index_ranges(if_bounds, orelse_bounds)
+        elif isinstance(s, (LoopIR.Assign, LoopIR.Reduce)):
+            # For Assign and Reduces, we are assuming that the RHS is computed before storing the
+            # result into the LHS, so we perform the checks in that order.
+
+            # First, the RHS
+            self.enter_scope()
+            super().do_e(s.rhs)
+            rhs_bounds = self.exit_scope()
+
+            if rhs_bounds is not None:
+                # We make no assumptions about the order of execution of the RHS
+                self.check_passed &= rhs_bounds.lo > rhs_bounds.hi - self.size
+                self.update_bounds(rhs_bounds, check_safety=True)
+
+            # Second, the LHS
+            if self.name == s.name:
+                bounds = index_range_analysis_wrapper(s.idx[self.dim], self.env)
+        else:
+            super().do_s(s)
+            bounds = self.exit_scope()
+
+        if bounds is not None:
+            self.update_bounds(bounds, check_safety=True)
+
+    def do_e(self, e):
+        if isinstance(e, LoopIR.Read) and e.name == self.name:
+            index_rng = index_range_analysis_wrapper(e.idx[self.dim], self.env)
+            self.update_bounds(index_rng)
+        elif isinstance(e, LoopIR.WindowExpr) and e.name == self.name:
+            w_access = e.idx[self.dim]
+            if isinstance(w_access, LoopIR.Interval):
+                lo_rng = index_range_analysis_wrapper(w_access.lo, self.env)
+                hi_rng = index_range_analysis_wrapper(w_access.hi, self.env)
+                self.update_bounds(lo_rng | hi_rng)
```

Comment:
Why is this not using `merge_index_ranges`?

Reply:
It doesn't have to because `index_range_analysis` will always return a valid IndexRange. `merge_index_ranges` is necessary if we don't know that the inputs might be `None` type.